tests/by-util/test_split.rs

```diff
@@ -1641,8 +1641,15 @@ fn test_round_robin() {
 #[test]
 #[cfg(any(target_os = "linux", target_os = "android"))]
 fn test_round_robin_limited_file_descriptors() {
-    new_ucmd!()
-        .args(&["-n", "r/40", "onehundredlines.txt"])
+    let scene = TestScenario::new(util_name!());
+
+    let mut shell_cmd = String::from("sleep 0.5 && "); // to ensure that the limit is applied
+    shell_cmd += &scene.bin_path.to_string_lossy();
+    shell_cmd += " split -n r/40 onehundredlines.txt";
+
+    scene
```

Comment:
Why not using the Rust function for this?
Seems overkill to start a shell for this

Reply:
The important part of this change is to add a `sleep 0.5` before execution of the unit under test (`split`).
This was for me the simplest way to do it.
Please tell me which rust function you refer to? It not clear to me.

Reply:
there is some strange error about not beeing able to link executable `sleep`. This is weird because in other sucessfull tests on the same virtual android device sleep is executed fine.

Reply:
Using this function https://doc.rust-lang.org/std/thread/fn.sleep.html

Reply:
I doubt that this is so easy to accomplish.
The simplest way without using a shell script would be to add this `std::thread::sleep()` at the very beginning of the main function of `split`. Maybe optionally activated by a cmd line arg. Is this what you mean?

Reply:
Why is sleeping required in the first place? The `limit()` method just pushes the restriction into a list, and `run_no_wait` applies it to the child *AFTER* the child is spawned. That's the wrong way around, as you correctly noticed. This bug affects two more tests that work with ulimit:
- `test_cat::test_closes_file_descriptors`
- `test_cp::test_closes_file_descriptors`

I think the "real" fix is to somehow wait in `std::process::Command::spawn` itself until `run_no_wait` is done applying all limits. Sadly, std does not support that. Maybe we could do a more involved fork-waitpipe-exec trick? Oh boy, that seems like a lot of effort.

How about doing this "sleep 0.5" in `UCommand::run_no_wait` instead? This way, you would fix all three tests, not just `test_round_robin_limited_file_descriptors`, and the code fixing the problem is closer to the place where the problem actually lies.

Reply:
I was heading for the "sleep 0.5" solution because it was the quickest way to solve the issue with the instability in this specific test. I still think that we should maybe first submit a workaround as the clean solution is probably much effort and/or will take some time as it involves changes in the std-library or other crates.

But in general I agree with you @BenWiederhake . The real issue is that "run_no_wait applies it to the child AFTER the child is spawned". And there seems to be no other solution for this because the function `prlimit` needs the PID of the process to be limited.
I was searching for a way to start a processes in a paused state directly from start. But I wasn't able to find something like this.

I also agree that some clean fix could be done in `std::process::Command::spawn`. E.g. one could do the `prlimit` after the `do_fork` on child side and before the `do_exec`. But thats deep in the std library and I guess a change there could take years?

So, the more pragmatic/realistic solution is to use the shell as a helper on child process side. Either with a sleep, or by calling the command line tool `ulimit` directly. I'm just a bit worried about the windows compatibility. Is it relevant here?

By the way, using the shell and the `pipe_in` feature for the commands in the shell solves the problem also wihtout sleep. I just realized this as I was more deeply investigate the code in `UCommand::run_no_wait`. The pipe_in happens after the `prlimit` is run. This way the shell anyway has to wait. But a shell is needed.

Another idea would be to use an other uutils tool like e.g. `env`. We could extend it with a parameter that puts the process into paused state directly after start and waits until we send a continue signal.

@BenWiederhake I did not understand how a sleep in the `UCommand::run_no_wait` could solve the issue? We need it on the side of the child process.

To sum it up: I would still propose to merge these changes (I'll remove the unneeded sleep 0.5 before) as a workaround for the not fixable problem of applying the prlimit after child process start.

After that, we probably want to remove the prlimit feature of our testframework as its impossible to make it stable without adaption on the `std::process::Command::spawn` and instead use a shell with `ulimit` command.

What do you think?

Reply:
Maybe there's some other crate that gives us more control over the process? There's a bit of an issue that it would need to be platform independent though or at least we need to be able to cast it to a normal `Command`.

Reply:
Or can we use `pre_exec` somehow? https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.pre_exec

Reply:
> I'm just a bit worried about the windows compatibility.

uutils only needs prlimit for the sake of testing, so I think it should be perfectly fine to simply not run these tests on Windows, for example.

Piping the command in to the shell might work; it sounds a bit fragile to me, but I can't put my finger on why.

> We could extend it with a parameter that puts the process into paused state directly after start

This would create an additional race: What if we accidentally send a SIGCONT before `env` has gone to sleep? Let's not do that.

Here's a new weird idea: Open a new pipe with `!CLO_EXEC`, and set a magic envvar that tells the program under test to attempt reading from it, before doing anything else. The parent process can then do all the necessary prlimit stuff, then close the pipe. No matter the order of pipe-read/pipe-close, the right thing happens. The special magic envvar could even be behind a `#[cfg(test)]` feature-gate or something, so that we don't leak this behavior in production.
This approach has the advantage of not requiring any unnecessary sleep, automatically waiting for however long the parent process needs, not introducing another race, not requiring to figure out any program paths, not requiring another shell that might accidentally interfere with signals/paths/timing/etc.

>  I did not understand how a sleep in the UCommand::run_no_wait could solve the issue?

Sorry, I meant moving your shell+sleep hack from the test into `UCommand::run_no_wait`.

Btw, I'm perfectly happy with the sleep as an intermediate solution. It works well enough, and removes a pressing issue that makes CI red.

Without sleep or any fix, the test is actually stronger, and in a useful way: It's nice that we can now deal with dynamically-changing limits.

I'm strongly in favor of keeping prlimit in the test framework: Prlimit is a necessary tool to test features.

I don't make the decisions here, but I think the pipe-into-shell and sleep solutions are good enough. If the cloexec-pipe-trick works as I'm hoping, I would love that as the longterm solution, but that's out of scope for this PR.

Thanks for taking care of the issue!

Reply:
@tertsdiepraam the `pre_exec` could be a solution. It seems to provide an injection point exactly where we need it. Apart from the `unsafe` trait it seems to be ideal. I will try if it works out like that.

Reply:
Oh, I didn't notice that. Yes, that sounds perfect! (And for tests, the `unsafe` should be absolutely fine I think.)